### PR TITLE
sc-hsm-tool: Add options to initialize with public key authentication

### PR DESCRIPTION
**Pull request by charredlot**
This PR rebases https://github.com/OpenSC/OpenSC/pull/1711 and reworks the CVC decoding per suggestions.

Tested on Nitrokey HSM 2 (SmartCard-HSM)

##### Checklist
- [x] Documentation is added or updated

### CODE REVIEW

File: src/libopensc/pkcs15-sc-hsm.c [hunk 1]
```diff
@@ -334,10 +418,332 @@ int sc_pkcs15emu_sc_hsm_decode_cvc(sc_pkcs15_card_t * p15card,
...
+	}
+	return r;
+}
+
+/*
+ * @param pka: will be overwritten, should be uninitialized or memset to 0
+ */
+int sc_pkcs15emu_sc_hsm_decode_pka(sc_pkcs15_card_t *p15card,
+											const u8 **buf, size_t *buflen,
+											sc_cvc_pka_t *pka)
```
Comment (frankmorgner) [inline src/libopensc/pkcs15-sc-hsm.c:676]: your indent setting looks very weired

File: src/tools/sc-hsm-tool.c [hunk 2]
```diff
@@ -1667,6 +1693,208 @@ static int unwrap_key(sc_card_t *card, int keyid, const char *inf, const char *p
...
+		r = -1;
+		goto err;
+	}
+	if (fread(pka, 1, sb.st_size, in) != (size_t)sb.st_size) {
+		perror(inf);
+		r = -1;
+		goto err;
+	}
+	fclose(in);
+	in = NULL;
```
Comment (frankmorgner) [inline src/tools/sc-hsm-tool.c:1848]: you may want to have a look at `fread_to_eof()`...

**Review by frankmorgner — CHANGES_REQUESTED** [review]
Thanks for lookng into this. The PR looks good at first glance, I've left some inline comments...

File: src/tools/sc-hsm-tool.c (hunk 2 above)
Comment (charredlot) [reply src/tools/sc-hsm-tool.c]: thanks, replaced with `fread_to_eof`

File: src/libopensc/pkcs15-sc-hsm.c (hunk 1 above)
Comment (charredlot) [reply src/libopensc/pkcs15-sc-hsm.c]: sorry not used to \t tabs :laughing: i changed the functions to overflow with a single indent

File: src/libopensc/pkcs15-sc-hsm.c [hunk 3]
```diff
@@ -334,10 +418,332 @@ int sc_pkcs15emu_sc_hsm_decode_cvc(sc_pkcs15_card_t * p15card,
...
+	r = sc_asn1_decode(p15card->card->ctx, format->seq, *buf, *buflen,
+			buf, buflen);
+	LOG_TEST_GOTO_ERR(card->ctx, r,
+		"Could not decode ASN.1 for public key file's old format");
+
+	r = 0;
+	/* fall-through */
+
+err:
+	if (format != NULL) {
```
[the remark this answers is not in the document]
Comment (charredlot) [reply src/libopensc/pkcs15-sc-hsm.c]: removed the checks for free

File: src/libopensc/pkcs15-sc-hsm.c [hunk 4]
```diff
@@ -334,10 +418,332 @@ int sc_pkcs15emu_sc_hsm_decode_cvc(sc_pkcs15_card_t * p15card,
...
+	format->seq[0] = format->data.asn1_public_key_req[0];
+	format->seq[1] = format->data.asn1_device_cvc.asn1_cvc[0];
+	format->seq[2] = format->data.asn1_dica_cvc.asn1_cvc[0];
+
+	r = sc_asn1_decode(p15card->card->ctx, format->seq, *buf, *buflen,
+			buf, buflen);
+	LOG_TEST_GOTO_ERR(card->ctx, r,
+		"Could not decode ASN.1 for public key file's old format");
+
+	r = 0;
```
[the remark this answers is not in the document]
Comment (charredlot) [reply src/libopensc/pkcs15-sc-hsm.c]: replaced 0 with `SC_SUCCESS`

File: src/libopensc/pkcs15-sc-hsm.c [hunk 5]
```diff
@@ -261,7 +281,90 @@ static int read_file(sc_pkcs15_card_t * p15card, u8 fid[2],
...
+ */
+static void sc_pkcs15emu_sc_hsm_format_asn1_cvcert(
+		struct sc_asn1_entry *asn1_cvcert, size_t asn1_cvcert_len,
+		struct sc_asn1_entry *asn1_cvc_body, size_t asn1_cvc_body_len,
+		struct sc_asn1_entry *asn1_cvc_pubkey, size_t asn1_cvc_pubkey_len,
+		sc_cvc_t *cvc)
+{
+	assert(asn1_cvc_pubkey_len >= C_ASN1_CVC_PUBKEY_SIZE);
+	assert(asn1_cvc_body_len >= C_ASN1_CVC_BODY_SIZE);
+	assert(asn1_cvcert_len >= C_ASN1_CVCERT_SIZE);
```
[the remark this answers is not in the document]
Comment (charredlot) [reply src/libopensc/pkcs15-sc-hsm.c]: changed to return an error that gets propagated through

File: src/libopensc/pkcs15-sc-hsm.c [hunk 6]
```diff
@@ -261,7 +281,90 @@ static int read_file(sc_pkcs15_card_t * p15card, u8 fid[2],
...
+ * @param asn1_authreq: unpopulated array with len matching c_asn1_authreq
+ * @param asn1_cvcert: already-initialized array matching c_asn1_cvcert
+ *
+ */
+static void sc_pkcs15emu_sc_hsm_format_asn1_req(
+		struct sc_asn1_entry *asn1_authreq, size_t asn1_authreq_len,
+		struct sc_asn1_entry *asn1_cvcert,
+		sc_cvc_t *cvc)
+{
+	assert(asn1_authreq_len >= C_ASN1_AUTHREQ_SIZE);
```
[the remark this answers is not in the document]
Comment (charredlot) [reply src/libopensc/pkcs15-sc-hsm.c]: changed to return an error

File: src/libopensc/card-sc-hsm.c [hunk 7]
```diff
@@ -1424,10 +1432,178 @@ static int sc_hsm_unwrap_key(sc_card_t *card, sc_cardctl_sc_hsm_wrapped_key_t *p
...
+	sc_format_apdu_ex(&apdu, 0x00, 0x54, 0x00, 0x00, NULL, 0, recvbuf, sizeof recvbuf);
+	apdu.cla = 0x80;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(ctx, r, "APDU transmit failed");
+
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
 	LOG_TEST_RET(ctx, r, "Check SW error");
 
+	print_public_key_auth_status(recvbuf);
```
[the remark this answers is not in the document]
Comment (charredlot) [reply src/libopensc/card-sc-hsm.c]: moved the printing to sc-hsm-tool

File: src/tools/Makefile.am [hunk 8]
```diff
@@ -81,7 +81,7 @@ openpgp_tool_SOURCES = openpgp-tool.c util.c
 openpgp_tool_LDADD = $(OPTIONAL_OPENSSL_LIBS)
 iasecc_tool_SOURCES = iasecc-tool.c util.c
 iasecc_tool_LDADD = $(OPTIONAL_OPENSSL_LIBS)
-sc_hsm_tool_SOURCES = sc-hsm-tool.c util.c
+sc_hsm_tool_SOURCES = sc-hsm-tool.c util.c fread_to_eof.c
```
Comment (frankmorgner) [inline src/tools/Makefile.am:84]: I think this also needs to be added in Makefile.mak

**Review by frankmorgner — CHANGES_REQUESTED** [review]

File: src/tools/Makefile.am (hunk 8 above)
Comment (charredlot) [reply src/tools/Makefile.am:84]: i attempted a change, but i'm not sure how to interpret this error

```
light.exe : error LGHT0103 : The system cannot find the file '..\src\tools\cardos-tool.exe' with type ''.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\WiX Toolset v3.11\\bin\light.exe"' : return code '0x67'
Stop.
Command exited with code 2
```

Comment (frankmorgner) [reply src/tools/Makefile.am:84]:
```
   Creating library pkcs11-spy.lib and object pkcs11-spy.exp
	if EXIST pkcs11-spy.dll.manifest mt -manifest pkcs11-spy.dll.manifest -outputresource:pkcs11-spy.dll;2
Makefile.mak(62) : fatal error U1001: syntax error : illegal character '^' in macro
Stop.
```

Comment (charredlot) [reply src/tools/Makefile.am:84]: oops thanks, that seems to have fixed it!

**Review by LOLSOR23 — APPROVED** [review]

**Review by Jakuje — APPROVED** [review]

**Comment by Jakuje** [comment]
Looks good to me. I think we should be fine merging this unless there would be some more comments.
It is a bit specific, but the code looks fine.

**Review by frankmorgner — APPROVED** [review]

**Comment by Staja** [comment]
Congrats on this, looking forward to testing this in 0.23.0

**Pull request merged**